Add an error state to the image block to allow upload errors to display

brentswisher opened the pull request:
## Description
Adds an error state to the image block to display errors that have occurred when a file fails to upload. This is an issue when a file is dropped into the editor or onto an empty image block, so the media library modal is not used.

Uses the existing noticeOperations methods for creating and displaying the errors passed back from the mediaUpload component. 

Fixes #10191  as well a similar issue I discovered when an image of an improper format (.psd) was used. 

## How has this been tested?
Open the editor and try to drag an image that is larger than the max file upload limit.
The error message passed back from the mediaUpload component should now be displayed in an empty image block.

## Types of changes
Adds a hasError variable to the state of the image block component. 
Sets this state based on the mediaUpload component passing back an error. 
If set to true, use noticeOperations to get and display the error messages.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

brentswisher committed:
Add a hasError state to the image block component that will display errors that have occurred when uploading invalid images

brentswisher committed:
Add a hasError state to the image block component that will display errors that have occurred when uploading invalid images

jorgefilipecosta commented on packages/block-library/src/image/edit.js:
```diff
@@ -72,23 +72,27 @@ class ImageEdit extends Component {
 					},
-					allowedType: 'image',
+					onError: ( message ) => {
+						this.setState( { hasError: true } );
```
I think if make a call here similar to:
```
						noticeOperations.createErrorNotice( message );
						this.props.setAttributes( {
							url: undefined,
							alt: undefined,
							id: undefined,
							caption: undefined,
						} );
```
We may be able to avoid the need for a state flag.

jorgefilipecosta reviewed (commented):
Hi @brentswisher thank you a lot for your contribution, things worked great in my tests 👍 
I left a possible suggestion for an improvement that may remove the need for a flag in the state.
Feel free to ping me if you need additional information!

brentswisher commented:
@jorgefilipecosta thanks for the feedback! I tried it the way you had suggested and that also displays the error, but there was a noticeable delay before the error message showed up. It seems almost like the file was complete uploaded before displaying the error? With the flag in the state, it shows the error almost instantaneously, which seems like maybe a better user experience. 

I don't entirely understand why your suggestion should slow the error displaying, so maybe I am doing something wrong. I modeled the hasError flag off of @noisysocks implementation of the file block, maybe he would have some insight?

brentswisher commented:
I'm on a bad connection right now but can upload some screen captures of the difference in error timing tomorrow.

brentswisher commented:
In looking into it a little bit more last night, it looks like the url gets set to a blob on the start of upload, and updated as the upload processes? I'm thinking it doesn't get set to undefined and stay until after it is finished uplaoding. With the state variable, it displays the error while the file is still getting uploaded, because it doesn't care what the url is set to. Again, very new to the gutenberg code so feel free to correct if I am mistaken.

Here is the upload process **without the state variable:**
![fun4c7aegf](https://user-images.githubusercontent.com/6653970/46212393-dde6aa80-c303-11e8-9f1f-8c32f0dcf89e.gif)


Here is the process **with the state variable**:
![mf6ltuxnmn](https://user-images.githubusercontent.com/6653970/46212027-05894300-c303-11e8-8252-37972762f855.gif)

brentswisher committed:
Merge branch 'master' into add/image-block-error-message

brentswisher committed:
Merge branch 'add/image-block-error-message' of github.com:brentswisher/gutenberg into add/image-block-error-message

jorgefilipecosta commented on packages/block-library/src/image/edit.js:
```diff
@@ -73,23 +73,27 @@ class ImageEdit extends Component {
 					},
-					allowedTypes: ALLOWED_MEDIA_TYPES,
+					onError: ( message ) => {
+						this.setState( { hasError: true } );
```
This seems a really complex problem if we apply a setAttributes here. The attributes are in fact changed in the store, but the components are not rerendered even if caching is disabled on the getBlocks selector. That's why there is a delay on the notice appearing when we use the setAttributes approach.

@noisysocks given that the state approach was also used on the file blocks do you have any insights on why setAttributes does not work for this cases?
I'm fine with the approach being proposed here my only fear is that the fact the setAttributes approach does not work well may be related to some problem that may have an impact in other areas.

Thank you @brentswisher for this PR and for the research you made.

brentswisher replied:
It looks like this isn't an issue in master anymore, wondering if #9264 may have fixed this issue as well and my PR is no longer needed?

noisysocks replied:
Using `setAttributes()` causes an undo level to be created.

brentswisher committed:
Merge branch 'master' into add/image-block-error-message

brentswisher committed:
Update state to use isEditing instead of hasError to display the error since it is already in master.

brentswisher commented:
The image block has changed a bit since I wrote the original bug fix but the error is still occurring. I merged master back into this branch so it is able to be merged and updated the code to simplify it a little bit. Since @talldan and @imath added the isEditing bit to the state in #9264, I rewrote my hasError addition to use that instead, does that clear up your concerns about this @jorgefilipecosta ?

talldan commented on packages/block-library/src/image/edit.js:
```diff
@@ -165,11 +169,9 @@ class ImageEdit extends Component {
 			} );
 			return;
 		}
-
```
I liked the whitespace! 😢

talldan replied:
Ah, ok! It's not the end of the world, but might be good to add it back if it was unintentional. There were a few others as well.

brentswisher replied:
Not a problem it should be all set now - thanks for pointing it out

brentswisher replied:
@talldan Replacing that whitespace triggered Travis CI and it failed for some unrelated reason 🤦‍♂️  
Is that something you can restart, or should I ask somewhere else?

talldan replied:
Hi @brentswisher - there were some minor issues after the 5.0 release with the CI jobs. Everything now seems to be passing so will merge this 👍

brentswisher replied:
Awesome, thanks!

talldan reviewed (approved):
Thanks for the contribution @brentswisher. 😄 

I gave it a test and confirmed it works. This is also consistent with the file block, so that's nice.

brentswisher committed:
Merge branch 'master' into add/image-block-error-message

brentswisher committed:
Add back whitespace that got accidentally removed earlier
